modules/network/src/main/java/org/apache/ignite/network/internal/netty/NettyClient.java

```diff
@@ -45,10 +51,16 @@
     @Nullable
     private volatile CompletableFuture<NettySender> clientFuture = null;
 
+    private CompletableFuture<Void> channelFuture = new CompletableFuture<>();
```

Comment:
1. Why do you need this future?
2. Javadoc is missing

Reply:
1. Basically, we had one future for client, `clientFuture`. It resolves not when the channel is open but when the handshake operation is complete. In case if we are stopping a client mid-handshake there is no way to close the channel immediately, that's why I added a future for "channel is open" event.